ChopChop/Storage/FirebaseDatabase.swift

```diff
@@ -0,0 +1,246 @@
+            })
+        })
+        .combineLatest
+        .map({
+            $0.compactMap({
+                $0
+            })
+        })
+        .eraseToAnyPublisher()
+    }
+
+    func fetchUserRating(userId: String) -> AnyPublisher<[UserRating], Error> {
+        db.collection(userPath).document(userId)
+            .getDocument()
+            .compactMap({
+                try? $0.data(as: User.self)?.ratings
+            })
+            .eraseToAnyPublisher()
+    }
+
+    func updateRecipeRating(userId: String, recipeId: String, newScore: RatingScore) {
+        let docRef = db.collection(recipePath).document(recipeId)
+        db.runTransction({ transaction -> Any? in
+            let document: DocumentSnapshot
+            do {
+                try document = transaction.getDocument(docRef)
+            } catch {
+                return nil
+            }
+            guard let recipeRatings = try? document.data(as: OnlineRecipeRecord.self)?.ratings else {
+                return nil
+            }
+
+            var newRatingList = [RecipeRating]()
+
+            for rating in recipeRatings {
+                if rating.userId == userId {
+                    newRatingList.append(RecipeRating(userId: userId, score: newScore))
+                } else {
+                    newRatingList.append(rating)
+                }
+            }
+
+            transaction.updateData(["ratings": newRatingList.map({ $0.toDict() })], forDocument: docRef)
+            return nil
+        })
+    }
+
+    func updateUserRating(userId: String, recipeId: String, newScore: RatingScore) {
+        let docRef = db.collection(userPath).document(userId)
+        db.runTransction({ transaction -> Any? in
+            let document: DocumentSnapshot
+            do {
+                try document = transaction.getDocument(docRef)
+            } catch {
+                return nil
+            }
+            guard let userRatings = try? document.data(as: User.self)?.ratings else {
+                return nil
+            }
+
+            var newRatingList = [UserRating]()
+
+            for rating in userRatings {
+                if rating.recipeId == recipeId {
+                    newRatingList.append(UserRating(recipeOnlineId: recipeId, score: newScore))
+                } else {
+                    newRatingList.append(rating)
+                }
+            }
+
+            transaction.updateData(["ratings": newRatingList.map({ $0.toDict() })], forDocument: docRef)
+            return nil
+        })
+    }
+
+    func fetchOnlineRecipeById(onlineRecipeId: String) -> AnyPublisher<OnlineRecipeRecord, Error> {
+
+        db.collection(recipePath).document(onlineRecipeId)
+            .publisher()
+            .compactMap({ try? $0.data(as: OnlineRecipeRecord.self) })
+            .eraseToAnyPublisher()
+    }
+
+    func fetchOnlineRecipeOnceById(onlineRecipeId: String) -> AnyPublisher<OnlineRecipeRecord, Error> {
```

Comment:
when is this used compared to ln 180

Reply:
Not sure the exact use case but this only fetches the recipe once and the publisher is completed so it doesnt listen to any changes unlike the one in 180